### ISSUE TITLE
Download links don't work when checking out as guest, even when Downloadable Product Options -> Order Item Status To Enable Downloads is set to pending

### ISSUE THREAD

**Issue by crankycyclops** [magento/magento2#20032]
### Preconditions (*)

1. Magento 2.2.7
2. PHP 7.1
3. MariaDB 10.2

### Steps to reproduce (*)

1. Make a downloadable product with shareable links
2. Set Admin -> Stores -> Configuration -> Catalog -> Downloadable Product Options -> Shareable to Yes
3. Set Admin -> Stores -> Configuration -> Catalog -> Downloadable Product Options -> Disable Guest Checkout if Cart Contains Downloadable Items to No.
4. Set Admin -> System -> Configuration -> Catalog -> Downloadable Product Options -> Order Item Status To Enable Downloads to 'Pending.'
5. Clear cache
6. Open a browser in incognito mode and add the downloadable product to your cart without logging in
7. Checkout as guest with a valid email address
8. Click on the link or links sent in the confirmation email

### Expected result (*)

1. Link should go straight to download

### Actual result (*)

1. User gets redirected to the login page, and they are not able to download the file
2. I had to manually update the entries in downloadable_link_purchased_item from 'pending' to 'available' before I could download them without logging in.

**Comment by magento-engcom-team** [comment]
Hi @crankycyclops. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@crankycyclops do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by crankycyclops** [comment]
I'm not sure how to test this on an automatically deployed vanilla instance, because the last time I tried for another ticket, I wasn't able to receive emails from it.

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97394`, `MAGETWO-97395` were created

**Comment by crankycyclops** [comment]
In case it helps with your debugging, I did some work on this to make a temporary fix for my own store, and in the process discovered that in module-downloadable/Observer/SetLinkStatusObserver.php on line 131 (for 2.2.7), the $linkPurchased collection is empty, which I don't think should be the case. As a result, none of the links have their status modified.

For now, I've patched SaveDownloadableOrderItemObserver.php, which works for my purposes as a stop gap, but is probably a bad solution.

```
diff -Naur a/vendor/magento/module-downloadable/Observer/SaveDownloadableOrderItemObserver.php b/vendor/magento/module-downloadable/Observer/SaveDownloadableOrderItemObserver.php
--- a/vendor/magento/module-downloadable/Observer/SaveDownloadableOrderItemObserver.php	2018-11-19 11:54:58.000000000 -0800
+++ b/vendor/magento/module-downloadable/Observer/SaveDownloadableOrderItemObserver.php	2019-01-02 02:50:46.096238415 -0800
@@ -123,6 +123,14 @@
                         ScopeInterface::SCOPE_STORE
                     );
                 $linkPurchased->setLinkSectionTitle($linkSectionTitle)->save();
+				/* Begin Hack */
+				// Stop using this as soon as Magento fixes this: https://github.com/magento/magento2/issues/20032
+				if (\Magento\Sales\Model\Order::STATE_COMPLETE == $orderItem->getOrder()->getState() && 'Guest' == $orderItem->getOrder()->getCustomerName()) {
+					$hackLinkState = \Magento\Downloadable\Model\Link\Purchased\Item::LINK_STATUS_AVAILABLE;
+				} else {
+					$hackLinkState = \Magento\Downloadable\Model\Link\Purchased\Item::LINK_STATUS_PENDING;
+				}
+				/* End Hack */
                 foreach ($linkIds as $linkId) {
                     if (isset($links[$linkId])) {
                         $linkPurchasedItem = $this->_createPurchasedItemModel()->setPurchasedId(
@@ -144,13 +152,16 @@
                             '+/=',
                             '-_,'
                         );
+						file_put_contents('/tmp/state', $orderItem->getOrder()->getState());
                         $numberOfDownloads = $links[$linkId]->getNumberOfDownloads() * $orderItem->getQtyOrdered();
                         $linkPurchasedItem->setLinkHash(
                             $linkHash
                         )->setNumberOfDownloadsBought(
                             $numberOfDownloads
                         )->setStatus(
-                            \Magento\Downloadable\Model\Link\Purchased\Item::LINK_STATUS_PENDING
+							/* Begin Hack */
+                            $hackLinkState
+							/* End Hack */
                         )->setCreatedAt(
                             $orderItem->getCreatedAt()
                         )->setUpdatedAt(
```

**Comment by magento-engcom-team** [comment]
Hi @GovindaSharma. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by GovindaSharma** [comment]
#MM19IN

**Comment by magento-engcom-team** [comment]
Hi @m-a-x-i-m. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if you want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by ajithkumar-maragathavel** [comment]
@magento I am working on this

**Comment by ajithkumar-maragathavel** [comment]
@magento Issue is not reproducible in the 2.4-develop branch.

**Comment by coderimus** [comment]
According to the issue history and current state of the 2.4 project, I close this issue.

**Comment by magento-engcom-team** [comment]
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment or reopen according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
the ticket if you are still facing this issue on the latest `2.x-develop` branch. Thank you for collaboration.
